File: src/test_ntt.cpp
#include <iostream>
#include <vector>

static const int64_t p = 998244353;
static const int n = 4;

inline int64_t mod(int64_t x) {
    int64_t r = x % p;
    return r < 0 ? r + p : r;
}

inline int64_t mod_pow(int64_t base, int64_t exp) {
    int64_t res = 1;
    base = mod(base);
    while (exp > 0) {
        if (exp & 1) res = (__int128(res) * base) % p;
        base = (__int128(base) * base) % p;
        exp >>= 1;
    }
    return res;
}

inline int64_t mod_inv(int64_t n) {
    return mod_pow(n, p - 2);
}

void bit_reverse(std::vector<int64_t>& a) {
    int deg = a.size();
    int j = 0;
    for (int i = 1; i < deg; ++i) {
        int bit = deg >> 1;
        while (j & bit) {
            j ^= bit;
            bit >>= 1;
        }
        j |= bit;
        if (i < j) std::swap(a[i], a[j]);
    }
}

void ntt(std::vector<int64_t>& a, bool invert) {
    int deg = a.size();
    for (int i = 0; i < deg; ++i) a[i] = mod(a[i]);
    bit_reverse(a);
    for (int len = 2; len <= deg; len <<= 1) {
        int64_t w_m = mod_pow(3, (p - 1) / len);
        if (invert) w_m = mod_inv(w_m);
        int half = len >> 1;
        std::vector<int64_t> w_pows(half);
        int64_t w = 1;
        for (int i = 0; i < half; ++i) {
            w_pows[i] = w;
            w = (__int128(w) * w_m) % p;
        }
        for (int start = 0; start < deg; start += len) {
            for (int i = 0; i < half; ++i) {
                int64_t u = a[start + i];
                int64_t v = (__int128(a[start + half + i]) * w_pows[i]) % p;
                a[start + i] = mod(u + v);
                a[start + half + i] = mod(u - v);
            }
        }
    }
    if (invert) {
        int64_t inv = mod_inv(deg);
        for (int i = 0; i < deg; ++i) {
            a[i] = (__int128(a[i]) * inv) % p;
        }
    }
}

int main() {
    std::vector<int64_t> a = {1, 2, 3, 4};
    ntt(a, false);
    std::cout << "NTT([1,2,3,4]) = ";
    for (auto x : a) std::cout << x << " ";
    std::cout << std::endl;
    ntt(a, true);
    std::cout << "INTT = ";
    for (auto x : a) std::cout << x << " ";
    std::cout << std::endl;
    return 0;
}

File: src/test_ntt.py
import numpy as np

p = 998244353

def bitrev(a):
    deg = len(a)
    j = 0
    a = a.copy()
    for i in range(1, deg):
        bit = deg >> 1
        while j & bit:
            j ^= bit
            bit >>= 1
        j |= bit
        if i < j:
            a[i], a[j] = a[j], a[i]
    return a

def ntt(a, invert=False):
    deg = len(a)
    a = bitrev(a % p)
    length = 2
    while length <= deg:
        w_m = pow(3, (p - 1) // length, p)
        if invert:
            w_m = pow(w_m, -1, p)
        half = length // 2
        w_pows = np.empty(half, dtype=np.int64)
        w = 1
        for i in range(half):
            w_pows[i] = w
            w = (w * w_m) % p
        for start in range(0, deg, length):
            lo = a[start:start + half].astype(np.int64)
            hi = (a[start + half:start + length].astype(np.int64) * w_pows) % p
            a[start:start + half] = (lo + hi) % p
            a[start + half:start + length] = (lo - hi) % p
        length <<= 1
    if invert:
        inv = pow(int(deg), -1, p)
        a = (a * inv) % p
    return a

a = np.array([1, 2, 3, 4], dtype=np.int64)
A = ntt(a)
print("Python NTT([1,2,3,4]) =", A)
print("Python INTT =", ntt(A, invert=True))
